In [2]:
import sys
value = int(sys.argv[1])
print("Valor usado:", value)

ValueError: invalid literal for int() with base 10: '-f'

### Reduction operation: the sum of the numbers in the range [0, a)

In [1]:
import time

def reduc_operation(a):
    """Compute the sum of the numbers in the range [0, a)."""
    x = 0
    for i in range(a):
        x += i
    return x

# Secuencial

value = 1000000

initialTime = time.time()
suma = reduc_operation(value)
finalTime = time.time()

print("Time taken by reduction operation:", (finalTime - initialTime), "seconds")

# Utilizando las operaciones mágicas de ipython
%timeit -r 2 reduc_operation(value)

print(f"\n \t Computing the sum of numbers in the range [0, value): {suma}\n")

Time taken by reduction operation: 0.037186622619628906 seconds
33.4 ms ± 646 μs per loop (mean ± std. dev. of 2 runs, 10 loops each)

 	 Computing the sum of numbers in the range [0, value): 499999500000



In [2]:
print("="*60)
print("Ejercicio a): Utilizando listas de python")
print("="*60)

#creamos una lista con range
print("\n1. Creando lista con list(range(value))...")
start = time.time()
lista = list(range(value))
tiempo_creacion_lista = time.time() - start
print(f"   Tiempo de creación: {tiempo_creacion_lista:.6f} segundos") #le ponemos un format para ajustar n de decimales

#Método 1: Suma con bucle for
def reduc_operation_lista_for(a):
    """Suma elementos de una lista usando bucle for."""
    x = 0
    for num in a:
        x += num
    return x

print("\n2. Suma con bucle for sobre lista")
initialTime = time.time()
suma_lista_for = reduc_operation_lista_for(lista)
finalTime = time.time()
tiempo_lista_for = finalTime - initialTime
print(f"   Resultado: {suma_lista_for}")
print(f"   Tiempo: {tiempo_lista_for:.6f} segundos")

# Con %timeit
print("\n   Usando %timeit:")
%timeit -r 2 reduc_operation_lista_for(lista)

# Método 2: Suma con función sum()
def reduc_operation_lista_sum(a):
    """Suma elementos de una lista usando sum()."""
    return sum(a)

print("\n3. Suma con función sum() sobre lista")
initialTime = time.time()
suma_lista_sum = reduc_operation_lista_sum(lista)
finalTime = time.time()
tiempo_lista_sum = finalTime - initialTime
print(f"   Resultado: {suma_lista_sum}")
print(f"   Tiempo: {tiempo_lista_sum:.6f} segundos")

# Con timeit
print("\n   Usando %timeit:")
%timeit -r 2 reduc_operation_lista_sum(lista)


Ejercicio a): Utilizando listas de python

1. Creando lista con list(range(value))...
   Tiempo de creación: 0.014899 segundos

2. Suma con bucle for sobre lista
   Resultado: 499999500000
   Tiempo: 0.025969 segundos

   Usando %timeit:
25.6 ms ± 82.1 μs per loop (mean ± std. dev. of 2 runs, 10 loops each)

3. Suma con función sum() sobre lista
   Resultado: 499999500000
   Tiempo: 0.006051 segundos

   Usando %timeit:
5.71 ms ± 7.72 μs per loop (mean ± std. dev. of 2 runs, 100 loops each)


In [3]:
import numpy as np

print("\n" + "="*60)
print("Ejercicio b: Utilizando arrays de numpy")
print("="*60)

# Convertir lista a array de numpy
print("\n1. Convirtiendo lista a array de numpy...")
start = time.time()
array_numpy = np.array(lista)
tiempo_conversion = time.time() - start
print(f"   Tiempo de conversión: {tiempo_conversion:.6f} segundos")

# Método 1: Suma con bucle for sobre array de numpy
def reduc_operation_numpy_for(a):
    """Suma elementos de un array numpy usando bucle for."""
    x = 0
    for num in a:
        x += num
    return x

print("\n2. Suma con bucle for sobre array nupmy")
initialTime = time.time()
suma_numpy_for = reduc_operation_numpy_for(array_numpy)
finalTime = time.time()
tiempo_numpy_for = finalTime - initialTime
print(f"   Resultado: {suma_numpy_for}")
print(f"   Tiempo: {tiempo_numpy_for:.6f} segundos")

#Con %timeit
print("\n   Usando %timeit:")
%timeit -r 2 reduc_operation_numpy_for(array_numpy)

# Método 2: Suma con np.sum()
def reduc_operation_numpy_sum(a):
    """Suma elementos de un array NumPy usando np.sum()."""
    return np.sum(a)

print("\n3. Suma con np.sum() sobre array de numpy")
initialTime = time.time()
suma_numpy_sum = reduc_operation_numpy_sum(array_numpy)
finalTime = time.time()
tiempo_numpy_sum = finalTime - initialTime
print(f"   Resultado: {suma_numpy_sum}")
print(f"   Tiempo: {tiempo_numpy_sum:.6f} segundos")

# Con timeit
print("\n   Usando %timeit:")
%timeit -r 2 reduc_operation_numpy_sum(array_numpy)



Ejercicio b: Utilizando arrays de numpy

1. Convirtiendo lista a array de numpy...
   Tiempo de conversión: 0.049647 segundos

2. Suma con bucle for sobre array nupmy
   Resultado: 499999500000
   Tiempo: 0.053073 segundos

   Usando %timeit:
54.3 ms ± 1.87 ms per loop (mean ± std. dev. of 2 runs, 10 loops each)

3. Suma con np.sum() sobre array de numpy
   Resultado: 499999500000
   Tiempo: 0.000389 segundos

   Usando %timeit:
152 μs ± 254 ns per loop (mean ± std. dev. of 2 runs, 10,000 loops each)


## Análisis de resultados

### Código original vs Lista

El código original (range + for) tarda 0.034091 segundos, mientras que trabajar directamente sobre una lista ya creada tarda 0.022599 segundos. Esto ocurre porque:

- En el código original, "range(value)" se genera dinámicamente dentro del bucle.
- Con la lista, los datos ya están en memoria y listos para usar
- Sin embargo, crear la lista consume memoria adicional

**Conclusión**: La lista es más rápida en el bucle, pero ocupa mucha memoria.

### Mejora con función sum()

La función "sum()" reduce notablemente el tiempo a 0.005943 segundos. Esto se debe a:

- Implementación en C
- Optimizaciones internas del lenguaje
- Menor número de operaciones intermedias

**Conclusión**: Ejecutar "sum()" directamente es más rápido que meterlo en una función.

### Arrays numpy

**Con bucle for**: El tiempo aumenta respecto a listas
- Causa: Cada acceso a un elemento requiere conversión python a numy

**Con np.sum()**: El tiempo se reduce a la mitad respecto a listas

**Conclusión final**: numpy solo es eficiente cuando usas sus funciones vectorizadas. Los bucles for son contraproducentes.

In [4]:
from numba import njit

In [5]:
from numba import njit

print("\n" + "="*60)
print("Ejercicio 3.3 a): Utilizando numba")
print("="*60)

# Aplicamos el  @njit a las funciones con un bucle for
@njit
def reduc_operation_numba_for(a):
    """Suma con bucle for optimizado con numba."""
    x = 0.0
    for num in a:
        x += num
    return x

# Aplicar @njit a la función con np.sum()
@njit
def reduc_operation_numba_sum(a):
    """Suma con np.sum() optimizado con Numba JIT."""
    return np.sum(a)

# Primera ejecución para compilar
# lo descartamos, si lo midiesemos aqui mide el t de ejecucion y de compilacion
print("\nCompilando funciones numba...")
_ = reduc_operation_numba_for(array_numpy)
_ = reduc_operation_numba_sum(array_numpy)
print("Compilacion completada.\n")

# medimos el tiempo con @njit -> bucle for
print("Suma con bucle for + @njit")
initialTime = time.time()
suma_numba_for = reduc_operation_numba_for(array_numpy)
finalTime = time.time()
tiempo_numba_for = finalTime - initialTime
print(f"Resultado: {int(suma_numba_for)}")
print(f"Tiempo: {tiempo_numba_for:.6f} segundos")

print("\nUsando %timeit:")
%timeit -r 2 reduc_operation_numba_for(array_numpy)

# Medir tiempo con @njit - np.sum()
print("\nSuma con np.sum() + @njit")
initialTime = time.time()
suma_numba_sum = reduc_operation_numba_sum(array_numpy)
finalTime = time.time()
tiempo_numba_sum = finalTime - initialTime
print(f"Resultado: {int(suma_numba_sum)}")
print(f"Tiempo: {tiempo_numba_sum:.6f} segundos")

print("\nUsando %timeit:")
%timeit -r 2 reduc_operation_numba_sum(array_numpy)


Ejercicio 3.3 a): Utilizando numba

Compilando funciones numba...
Compilacion completada.

Suma con bucle for + @njit
Resultado: 499999500000
Tiempo: 0.001064 segundos

Usando %timeit:
940 μs ± 1.17 μs per loop (mean ± std. dev. of 2 runs, 1,000 loops each)

Suma con np.sum() + @njit
Resultado: 499999500000
Tiempo: 0.000238 segundos

Usando %timeit:
122 μs ± 14.6 ns per loop (mean ± std. dev. of 2 runs, 10,000 loops each)


## Resultados con numba

### Comparación de tiempos

- Numpy + for (sin @njit) -> 54.3 ms **(1x)**
- Numpy + for + @njit -> 940 μs **(57.7x)** |
- np.sum() (sin @njit) -> 152 μs **(357x)**
- np.sum() + @njit -> 122 μs **(445x)**

### Análisis

**Bucle for con @njit:**
- Sin Numba: 54.3 ms
- Con Numba: 940 μs
- **Mejora: aprox 58x más rápido**

**np.sum() con @njit:**
- Sin Numba: 152 μs  
- Con Numba: 122 μs
- **Mejora: aprox 1.25x más rápido**

Como observamos en el análisis, numba acelera enormemente los bucles for sobre el array de numpy. Además, aunque en menor medida, también consigue acelerar la funcion np.sum().

## Ejecución en Cola SLURM (ampere)

### Resultados con N = 10⁷ (10 millones)
-------------------
Job ID: 12657
Nodo: ampere
Partition: hpc-bio-ampere
Fecha: Sun Dec 14 02:21:35 CET 2025
--------------------

Ejecutando notebook con value = 10000000

Valor usado: 10000000
Time taken by reduction operation: 0.034571170806884766 seconds
33 ms ± 24.5 μs per loop (mean ± std. dev. of 2 runs, 10 loops each)

 	 Computing the sum of numbers in the range [0, value): 499999500000

============================================================
Ejercicio a): Utilizando listas de python
============================================================

1. Creando lista con list(range(value))...
   Tiempo de creación: 0.016746 segundos

2. Suma con bucle for sobre lista
   Resultado: 499999500000
   Tiempo: 0.026693 segundos

   Usando %timeit:
26.6 ms ± 23.7 μs per loop (mean ± std. dev. of 2 runs, 10 loops each)

3. Suma con función sum() sobre lista
   Resultado: 499999500000
   Tiempo: 0.006164 segundos

   Usando %timeit:
6.12 ms ± 3.62 μs per loop (mean ± std. dev. of 2 runs, 100 loops each)

============================================================
Ejercicio b: Utilizando arrays de numpy
============================================================

1. Convirtiendo lista a array de numpy...
   Tiempo de conversión: 0.041772 segundos

2. Suma con bucle for sobre array nupmy
   Resultado: 499999500000
   Tiempo: 0.061532 segundos

   Usando %timeit:
59.6 ms ± 454 μs per loop (mean ± std. dev. of 2 runs, 10 loops each)

3. Suma con np.sum() sobre array de numpy
   Resultado: 499999500000
   Tiempo: 0.000351 segundos

   Usando %timeit:
225 μs ± 233 ns per loop (mean ± std. dev. of 2 runs, 1,000 loops each)

============================================================
Ejercicio 3.3 a): Utilizando numba
============================================================

Compilando funciones numba...
Compilacion completada.

Suma con bucle for + @njit
Resultado: 499999500000
Tiempo: 0.001140 segundos

Usando %timeit:
1.01 ms ± 53.7 ns per loop (mean ± std. dev. of 2 runs, 1,000 loops each)

Suma con np.sum() + @njit
Resultado: 499999500000
Tiempo: 0.000230 segundos

Usando %timeit:
172 μs ± 13.9 ns per loop (mean ± std. dev. of 2 runs, 10,000 loops each)

--------------------
Job finalizado: Sun Dec 14 02:21:51 CET 2025
--------------------


### Resultados con N = 10⁸ (100 millones)
-------------------
Job ID: 12658
Nodo: ampere
Partition: hpc-bio-ampere
Fecha: Sun Dec 14 02:23:18 CET 2025
--------------------

Ejecutando notebook con value = 100000000

Valor usado: 100000000
Time taken by reduction operation: 0.03716588020324707 seconds
34.3 ms ± 133 μs per loop (mean ± std. dev. of 2 runs, 10 loops each)

 	 Computing the sum of numbers in the range [0, value): 499999500000

============================================================
Ejercicio a): Utilizando listas de python
============================================================

1. Creando lista con list(range(value))...
   Tiempo de creación: 0.015944 segundos

2. Suma con bucle for sobre lista
   Resultado: 499999500000
   Tiempo: 0.024698 segundos

   Usando %timeit:
24.3 ms ± 14.8 μs per loop (mean ± std. dev. of 2 runs, 10 loops each)

3. Suma con función sum() sobre lista
   Resultado: 499999500000
   Tiempo: 0.006169 segundos

   Usando %timeit:
6.12 ms ± 440 ns per loop (mean ± std. dev. of 2 runs, 100 loops each)

============================================================
Ejercicio b: Utilizando arrays de numpy
============================================================

1. Convirtiendo lista a array de numpy...
   Tiempo de conversión: 0.041644 segundos

2. Suma con bucle for sobre array nupmy
   Resultado: 499999500000
   Tiempo: 0.061622 segundos

   Usando %timeit:
60.6 ms ± 320 μs per loop (mean ± std. dev. of 2 runs, 10 loops each)

3. Suma con np.sum() sobre array de numpy
   Resultado: 499999500000
   Tiempo: 0.000354 segundos

   Usando %timeit:
247 μs ± 133 ns per loop (mean ± std. dev. of 2 runs, 1,000 loops each)

============================================================
Ejercicio 3.3 a): Utilizando numba
============================================================

Compilando funciones numba...
Compilacion completada.

Suma con bucle for + @njit
Resultado: 499999500000
Tiempo: 0.001142 segundos

Usando %timeit:
1.02 ms ± 210 ns per loop (mean ± std. dev. of 2 runs, 1,000 loops each)

Suma con np.sum() + @njit
Resultado: 499999500000
Tiempo: 0.000239 segundos

Usando %timeit:
192 μs ± 11.1 ns per loop (mean ± std. dev. of 2 runs, 10,000 loops each)

--------------------
Job finalizado: Sun Dec 14 02:23:35 CET 2025
--------------------

### Observaciones

- **Cola utilizada**: ampere
- **Método más eficiente**: numba con njit y np.sum() -> 0.000230 y 0.000239 segundos, respectivamente